ops-catalog/catalog-stats.ts

```diff
@@ -0,0 +1,174 @@
+import { IDerivation, Document, Register, ByGrainSource } from 'flow/ops/TENANT/catalog-stats';
+
+// Implementation for derivation template-common.flow.yaml#/collections/ops~1TENANT~1catalog-stats/derivation.
+export class Derivation implements IDerivation {
+    byGrainPublish(source: ByGrainSource, _register: Register, _previous: Register): Document[] {
+        const ts = new Date(source.ts);
+        const grains = grainsFromTS(ts);
+
+        const taskDocs = mapStatsToDocsByGrain(grains, taskStats(source)).map((doc) => ({
+            ...doc,
+            // For documents generated specific to this task, retain the detailed information about
+            // the task itself.
+            taskStats: {
+                capture: source.capture,
+                derive: source.derive,
+                materialize: source.materialize,
+            },
+        }));
+
+        // Documents generated for collections involved in this task will not have associated
+        // detailed task information. If the collection is a derivation, that will be accounted for
+        // above.
+        const collectionDocs = mapStatsToDocsByGrain(grains, collectionStats(source));
+
+        return [...taskDocs, ...collectionDocs];
+    }
+}
+
+type TimeGrain = {
+    grain: Document['grain'];
+    ts: string;
+};
+
+const grainsFromTS = (ts: Date): TimeGrain[] => {
+    ts.setUTCMilliseconds(0);
+    ts.setUTCSeconds(0);
+    ts.setUTCMinutes(0);
+
+    const hourlyTS = ts.toISOString();
+    ts.setUTCHours(0);
+    const dailyTS = ts.toISOString();
+    ts.setUTCDate(1);
+    const monthlyTS = ts.toISOString();
+
+    return [
+        {
+            grain: 'hourly' as Document['grain'],
+            ts: hourlyTS,
+        },
+        {
+            grain: 'daily' as Document['grain'],
+            ts: dailyTS,
+        },
+        {
+            grain: 'monthly' as Document['grain'],
+            ts: monthlyTS,
+        },
+    ];
+};
+
+type StatsData = {
+    [k: string]: Document['statsSummary'];
+};
+
+const newDocumentStats = (): Document['statsSummary'] => ({
+    writtenBy: {
```

Comment:
Is it possible to make this more dynamic? I'm concerned about the space and processing time for all of these zero values.

There's a related issue we'll need to tackle before long, where the stats documents we currently write out have an awful lot of explicit zero-value counters.

I'm guessing this could be a little tricky because of the materialization -- we don't want the materialized columns to have NULL values, as that makes roll-ups weird. Tangentially, it would be nice if the were able to project a `default` json schema annotation keyword all the way into a Postgres column default... (and we do have the `default` [available via inference](https://github.com/estuary/flow/blob/master/go/protocols/flow/flow.proto#L230-L231))

It's okay to punt on this if it's a rabbit hole.

Reply:
We're already creating the table with the schema we want so we can set `default 0`, but the materialization will still try to insert `null` values if the field isn't present in the document. If the column has a `not null` constraint it will fail at this, or it will successfully insert a `null` if there isn't a `not null`, even with a `default 0`. There's probably a way to enhance the connector to enable a configuration where instead of inserting `null` for a column of a new record, it would instead insert whatever the default value was. Perhaps that is what you were getting at, beyond just the column default.

Would it really be that bad to have `null` values in the materialized table? Queries could perhaps do a `coalesce`, or just deal with a `null`.